### ISSUE TITLE
BatchNormBackward crashes on 3D input with IndexOutOfRangeException (missing BatchNormBackward3D path)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#233]
## Summary

`CpuEngine.BatchNormBackward<T>` does not handle 3D input. The forward
path (`CpuEngine.BatchNorm<T>`) routes 3D tensors to a dedicated
`BatchNorm3D` helper that treats `_shape[0]` as **channels** and
`_shape[1..2]` as spatial dims, and records
`mean`/`variance` of length `C = _shape[0]`. The backward path has
no matching `BatchNormBackward3D` — 3D tensors fall through to the
2D `[batch, features]` code, which assumes `_shape[1]` is the
feature axis and indexes `meanData[f]` for `f ∈ [0, _shape[1])`.

Whenever `_shape[1] > _shape[0]` on a 3D input, `meanData` (length
`_shape[0]`) is indexed out of bounds → `System.IndexOutOfRangeException`.

## Versions

Reproduced from `main`. The code is identical between `v0.46.1` and
`v0.52.0` — the gap was introduced earlier and is still present on
latest.

## Repro

Any `Tensor<T>` of rank 3 with `_shape[1] > _shape[0]` fed through
the tape-recorded `BatchNorm` forward then backpropped:

```csharp
var input = new Tensor<double>(new[] { 1, 8, 24 });
var gamma = new Tensor<double>(new[] { 1 });
var beta  = new Tensor<double>(new[] { 1 });
using var tape = new GradientTape<double>();
var y = engine.BatchNorm(input, gamma, beta, 1e-5,
                         out var mean, out var variance);
var loss = engine.TensorSum(y);
tape.ComputeGradients(loss); // System.IndexOutOfRangeException
```

Downstream: this blocks training for ~17 models in the consuming
`AiDotNet` Finance smoke suite (`FlowState`, `GPT4TS`, `Kronos`,
`LLMTime`, `MOMENT`, `SimMTM`, `Sundial`, `TEST`, `TFC`, `TOTO`,
`TS2Vec`, `TimeBridge`, `TimeMAE`, `TimeMoE`, `TinyTimeMixers`,
`VisionTS`, `YingLong`) when their default layer stacks include
`BatchNormalizationLayer` and the test-harness inputs come through
as rank-3 `[batch, seq, features]`.

Stack trace from the consuming repo:

```
System.IndexOutOfRangeException : Index was outside the bounds of the array.
   at AiDotNet.Tensors.Engines.Autodiff.BackwardFunctions`1.BatchNormBackward(...)
   at AiDotNet.Tensors.Engines.Autodiff.CompiledDelegateChain`1.Execute(...)
   at AiDotNet.Tensors.Engines.Autodiff.GradientTape`1.ComputeGradientsViaGraph(...)
   at AiDotNet.Tensors.Engines.Autodiff.GradientTape`1.ComputeGradients(...)
```

## Root cause

`src/AiDotNet.Tensors/Engines/CpuEngine.cs`:

- **Forward** (`BatchNorm<T>`, line ~14718): routes `input._shape.Length == 3`
  to `BatchNorm3D`, which stores `mean`/`variance` of length `_shape[0]`
  (channels).
- **Backward** (`BatchNormBackward<T>`, line ~15216): only branches on
  `_shape.Length == 4`. All other ranks use the 2D path which reads
  `input._shape[0]` as `batch` and `input._shape[1]` as `features`,
  and indexes `meanData[f]` / `varData[f]` / `gammaData[f]` /
  `gradGammaData[f]` for `f ∈ [0, features)`. Those arrays have
  length `_shape[0]`, not `_shape[1]` — out-of-bounds whenever
  `_shape[1] > _shape[0]`.

## Proposed fix

Add a `BatchNormBackward3D<T>` helper that mirrors `BatchNorm3D`'s
channels-first layout (channels = `_shape[0]`, spatial = `_shape[1] *
_shape[2]`), and dispatch on `input._shape.Length == 3` alongside the
existing `== 4` branch in `BatchNormBackward<T>`. The gradient formula
is the same — accumulate `gGamma` / `gBeta` / `sumGrad` / `sumGradX`
per channel across the spatial axis instead of across the batch axis.

Happy to send a PR if this looks right; flagging the issue first so
ownership is clear.

---

Found while working through the Finance smoke-suite cleanup in
`ooples/AiDotNet` PR #1177.
